### DISCUSSION TITLE
ApiKey junto com Roles

### DISCUSSION THREAD

**Discussion by mathfrance** [balta-io/forum#266] [Tira Dúvidas] [answered]
Na aula de Autenticação e Autorização de ASP.NET 6 nos foi mostrada essas duas formas de autorização, gostaria de saber se é possível e como faria para o mesmo método autorizar pela Apikey ou caso o usuário tenha o role de admin?

**Comment by andrebaltieri** [comment, marked as answer]
Bom dia, @mathfrance, como vai?

É possível, seria algo como: Não veio ApiKey, eu valido por roles? Tem algum cenário para exemplificar a necessidade do uso de ambos juntos?

Lembrando que os atributos são executados na ordem, então:
```csharp
[ApiKey] // Executa primeiro
[Authorize()] // Executa depois
public IActionResult....

```

**Reply by mathfrance** [reply]
Então, eu deixei nessa ordem e até invertida, mas eu só recebi "ApiKey não encontrada", mesmo passando o Token. (Tanto que quando eu removo a "[ApiKey]" ele autoriza).

P.s: o método é um post, não sei se teria algum impacto.

Quanto ao cenário seria um método para fazer setup do Role = admin para um email passado por post. Podendo ser executado por um user admin ou alguém que tivesse acesso a ApiKey.

**Reply by andrebaltieri** [reply]
Então, mas aí você precisa mudar o código do ApiKey para verificar se existe um Role (Ou algo similar).
Os ApiKeys são para cenários onde vamos autenticar de servidor para servidor ou de um Job para servidor, diferente do Authorize. Por isto nunca pensei em usá-los juntos.

**Reply by mathfrance** [reply]
Entendi. Só queria deixar um cenário que aplicasse essa forma de autorização, daí acabei caindo nesse. Obrigado Balta.

**Reply by andrebaltieri** [reply]
Beleza @mathfrance, se tiver algum exemplo manda aí... mas é que normalmente é como citei acima... abraços!

**Reply by oimarcella** [reply]
E se tiver uma Apikey específica para cada tipo de role? é uma ideoa...

**Comment by ghost** [comment]
Em "Implementando um ApiKey" você disse que o cenário é de servidor para servidor ou job para servidor. Mas se as Apis ou métodos pudessem ser acessados, também, por um usuário usando login e senha, utilizando suas Roles? 

Tudo isso também funciona para Minimal Apis? 

Essa abordagem de Attributes funciona como um Interceptor de requisições?

**Reply by andrebaltieri** [reply]
Funcionam, mas normalmente não com atributos e sim com Policies nas Minimal APIs.
```csharp
using Microsoft.Identity.Web;

var builder = WebApplication.CreateBuilder(args);

builder.Services.AddAuthorizationBuilder()
  .AddPolicy("admin_greetings", policy =>
        policy
            .RequireRole("admin")
            .RequireClaim("scope", "greetings_api"));

var app = builder.Build();

app.MapGet("/hello", () => "Hello world!")
  .RequireAuthorization("admin_greetings");

app.Run();
```
https://learn.microsoft.com/en-us/aspnet/core/fundamentals/minimal-apis/security?view=aspnetcore-9.0
